### PR TITLE
Update and fix Linphone

### PR DESCRIPTION
**Pull request by lorenz**
###### Motivation for this change
Linphone in Nix is currently quite broken, it doesn't even launch without fixes since at least a few months. See #144577.
Additionally lots of features were broken, including the desktop icon, ringtones and video calling.

###### Things done
This change bumps Linphone to 4.4.1 and fixes a bunch of broken things. I've successfully tested ICE-enabled encrypted video calls and all starter-related things now look fine (including the autostarter). This contains lots of library updates, but most of them are just libraries by the authors of Linphone itself.

Thanks to @ThibaultLemaire who co-authored this and fixed issues with Linphone's own account system. 

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by austinbutler** [comment]
Wish I had seen your PR before I started messing with this! Anyways, in case it's helpful here's my branch where everything is up to date and builds except `linphone` itself, in case it's helpful: https://github.com/austinbutler/nixpkgs/tree/liblinphone-5.

File: pkgs/development/libraries/liblinphone/use-normal-jsoncpp.patch [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
+index 5cd87fe9a..822b2151e 100644
+--- a/src/CMakeLists.txt
++++ b/src/CMakeLists.txt
+@@ -34,7 +34,7 @@ if (ENABLE_FLEXIAPI)
+ 	if (XCODE)
+ 		list(APPEND LINK_LIBS jsoncpp_static)
+ 	else()
+-		list(APPEND LINK_LIBS jsoncpp_object)
++		list(APPEND LINK_LIBS jsoncpp)
```
Comment (austinbutler) [inline pkgs/development/libraries/liblinphone/use-normal-jsoncpp.patch:20]: The other alternative is re-enabling `BUILD_OBJECT_LIBS` in `jsoncpp`: https://github.com/austinbutler/nixpkgs/blob/8726b4987c4160f9f039ac475a88cad8d0eba1af/pkgs/development/libraries/jsoncpp/default.nix#L38

It was turned off in https://github.com/NixOS/nixpkgs/pull/102379, but I wonder if it's worth having compatibility issues that require patches like this.

Comment (lorenz) [reply pkgs/development/libraries/liblinphone/use-normal-jsoncpp.patch:20]: I know, I initially flipped that. But what this does is statically link object files from another package which is not really wanted here.

**Comment by lorenz** [comment]
I do also have a branch with everything updated to 5.1.12, but since Linphone itself does not build on it even on master I just kept that locally.

File: pkgs/development/libraries/belcard/default.nix [hunk 2]
```diff
@@ -7,7 +7,7 @@
 
 stdenv.mkDerivation rec {
   pname = "belcard";
-  version = "5.0.55";
+  version = "5.0.66";
```
Comment (7c6f434c) [inline pkgs/development/libraries/belcard/default.nix:10]: Shouldn't there be a mtching hash bump?

**Comment by ThibaultLemaire** [comment]
Hi! Linphone developer here. Thank you so much for your work on this! I have tested some features functional but am still getting segfaults with many other things.

Do you know how I could build a debug version (ideally also with sanitizer) so I could investigate and fix the issues?

Tested functional:
- unregistered video & audio call
- sending unencrypted chat message

Segfaulting:
- connecting to a sip.linphone.org account
- receiving a message (that one is *a bit* of a problem)

---
Context:
I have been using NixOS personally for a few years, and recently got hired as a developer on the Linphone project. I would now want to use NixOS on my work computer, and obtained permission from my employer to work on fixing Linphone on NixOS (though as a low priority task).

I have built the Linphone desktop app on Ubuntu as part of my work, and have written some basic nix-shell environments for myself, but don't know the tips and tricks of full-blown nixpkgs derivations (yet).

**Comment by lorenz** [comment]
I'm only using Linphone as a normal SIP softphone so I haven't tested your account handling. Can you compile with debug symbols and post the stack trace?

File: pkgs/development/libraries/belcard/default.nix (hunk 2 above)
Comment (ThibaultLemaire) [reply pkgs/development/libraries/belcard/default.nix:10]: Both tags are on the same commit, so no hash bumps. This is more of a semantic change.

**Comment by ThibaultLemaire** [comment]
I have updated linphone and its dependencies to 4.4.1 (which is the latest stable as of writing), though that didn't fix the issue.
I also managed to build with debug symbols and pinned one of the segfaults I'm experiencing as coming from `lime` (**Li**nphone **m**edia **e**ncryption) the library we use for secure communications.

I don't know what's the problem yet and will keep investigating.

**Comment by ThibaultLemaire** [comment]
I fixed this particular segfault: `bctoolbox` was missing `decaf` (an elliptic curve library) which in turn made `lime` unusable.

Registration on sip.linphone.org and encrypted audio/video calls seem to be working now. I still have a segfault related to chat messages.

I also noticed that we (nixpkgs) use upstream `libmatroska` and `soci`, which apparently builds, but Linphone has vendored versions of those libraries with differences that I have yet to investigate, but could prove to be an issue.

**Comment by lorenz** [comment]
@ThibaultLemaire Regarding `decaf`: Assuming we're talking about https://github.com/krionbsd/libdecaf, why is Linphone using this library? It seems abandoned which is not great for a cryptography library. The x448 and x25519 ECDH primitives are in OpenSSL/libcrypto as far as I know which you link with anyways.

**Comment by ThibaultLemaire** [comment]
What you linked seems to be a shallow copy of the original library which resides [here](https://sourceforge.net/p/ed448goldilocks/code/ci/master/tree/) I believe, and has seen changes last year.

As to why we're using crypto primitives from decaf if the same are available in OpenSSL, that is a good question... That I shall have to ask my colleagues, as I'm not a security expert, and this work was prior to my enrollment.

Also, let us agree that this is orthogonal to making Linphone run on NixOS, and would have to be fixed (if at all) upstream in an upcoming release of Linphone.

**Comment by ThibaultLemaire** [comment]
Ok, so, we **don't** link with OpenSSL (you can check the `buildInputs` of the various Belledonne Communications libraries in this very PR) because it's too heavy a dependency for mobile apps. So for cryptography we use `mbedtls` + `decaf` for elliptic curves.

**Comment by 7c6f434c** [comment]
Well, either @lorenz does something with your changes (and with merge conflicts here), or you open another PR, I do not have strong opinions where exactly to merge the same changes…

**Comment by ThibaultLemaire** [comment]
The rebase went well. Thanks for your review, @lorenz, I'll be waiting for your decision on which version (4.3.2 or 4.4.1) to push forward and how (this PR, another one?). (no hurry though)

File: pkgs/development/libraries/ortp/default.nix [hunk 3]
```diff
@@ -6,15 +6,15 @@
 
 stdenv.mkDerivation rec {
   pname = "ortp";
-  version = "4.5.1";
+  version = "5.0.66";
```
Comment (sikmir) [inline pkgs/development/libraries/ortp/default.nix:9]: 5.1.3 is the latest one as for now.

Comment (ThibaultLemaire) [reply pkgs/development/libraries/ortp/default.nix:9]: 5.1.12 actually (though that's [the same commit](https://gitlab.linphone.org/BC/public/ortp/-/commit/23439fba3b800c2fdb105af748ca018f51146d57)). I have made [that change in my update to 4.4.1](https://github.com/ThibaultLemaire/nixpkgs/commit/e57e1465d90f03082a52e38a73ce477aa1fb0636#diff-a0c8d04f47ef57ec3f17101235177d5624be63fd87d4124dc4ebbff874f0f8f8R10).

However linphone 4.3.2 has ortp pinned to [13731952b44192eb9be3e517a370df05c449dd28](https://gitlab.linphone.org/BC/public/ortp/-/commit/13731952b44192eb9be3e517a370df05c449dd28) so 5.0.66 is adequate here.

Comment (sikmir) [reply pkgs/development/libraries/ortp/default.nix:9]: I see, thanks.

**Comment by lorenz** [comment]
@ThibaultLemaire I integrated your work and did some more cleanup mainly related to naming and flags. Could you test what's currently in this PR and report back? If everything works we can merge this.

**Comment by ThibaultLemaire** [comment]
@lorenz It builds and works fine! (functionality tested: basic&encrypted chat, encrypted audio&video call)

I don't know what you mean by "cleanup [...] related to naming", but I see you decided to use `-DENABLE_STRICT=NO` to disable warnings turning to errors instead of disabling warnings individually. I'm curious, what's the reasoning behind this, ease of maintenance? What are the pros and cons? (This a non-blocker, I don't have an opinion on the topic.)

Also, have you seen my dependency cleanup commit? https://github.com/ThibaultLemaire/nixpkgs/commit/3dccfe356c86834ea02d4fbc50df8d4b52f2a476
As mentioned in the message I could have missed some things, but to the extent I have tested, everything works the same.

Additionally, I saw release notes mentioned somewhere in the contribution guidelines, should we update those as part of this PR?

**Comment by lorenz** [comment]
Thanks for testing! The reason I disabled `ENABLE_STRICT` is because it enables `-Werror` which IMO should never be used as part of a normal build, only as part of a CI/lint pass on the upstream project. Having it enabled essentially guarantees that a project only builds using a specific version of a specific compiler as any future versions or different compilers will probably have better/more warnings, failing the build. Since this is a downstream build the maintainers here aren't in a position to fix the warnings anyways.

I saw the dependency cleanup commit, but I somehow missed integrating it. I'll fix it.

Not sure if we should bother with release notes as this is essentially just a big update. But I'll let the reviewers be the judge of that.

**Comment by 7c6f434c** [comment]
I think for the users who use Linphone and switch from whatever revision of Nixpkgs to post-this-PR there is nothing to be extra careful about. And while Linphone is a good and important piece of software, I see no indication of >10% of Nixpkgs users using it or something like that. So I see neither «high risk or intentional breakage for the chosen few» nor «large impact for a huge number», hence there is clearly (to me) no release notes obligation.

**Comment by yorickvP** [comment]
This commit changes the version number of belle-sip from 4.5.14 to linphone-4.4.1 (which is actually around 5.1.30). However, CVE's use belle-sip's own version number, so it might be good to use that.

**Comment by ThibaultLemaire** [comment]
"around" is the problem here. The official release 4.4.1 of Linphone Desktop references a commit in belle-sip that is exactly 1 commit after tag 5.1.12 of belle-sip and 2 commits behind the next tag, which is 5.1.31 (or 5.1.32). So the belle-sip version we currently use in nixpkgs is *not* an official release of belle-sip (and therefore does not have a tag).

Your point is valid and should be taken into account for the next time I (or someone else) updates Linphone. In the meantime, I will try to sort releases with my colleagues at Belledonne Communications to have consistent versioning.

**Pull request merged**